### PR TITLE
Support pagination to fetch _all_ Bitbucket branches

### PR DESCRIPTION
**Pull request by jankeromnes**
## Description

<details>
<summary>Summary generated by Copilot</summary>

### <samp>🤖 Generated by Copilot at f818630</samp>

Add pagination support to `listBranches` method of `BitbucketRepositoryProvider`. This allows fetching all the branches for a Bitbucket repository when starting a workspace.

</details>

## Related Issue(s)

Fixes https://linear.app/gitpod/issue/EXP-3/bitbucketorg-support-api-pagination-when-listing-branches

## How to test

1. ?

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - jx-paginatedfeca2734</li>
	<li><b>🔗 URL</b> - <a href="https://jx-paginatedfeca2734.preview.gitpod-dev.com/workspaces" target="_blank">jx-paginatedfeca2734.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
	<li><b>📦 Version</b> - jx-paginate-bitbucket-branches-gha.15806</li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment / Integration Tests</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [x] /werft with-large-vm
- [x] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`. If enabled, `with-preview` and `with-large-vm` will be enabled.
- [ ] with-monitoring
</details>

/hold

### CODE REVIEW

**Comment by jankeromnes** [comment]
FYI, reviewing while ignoring whitespace changes is much easier here:

https://github.com/gitpod-io/gitpod/pull/18563/files?w=1

**Comment by loujaybee** [comment]
Thanks, @jankeromnes for raising a fix ! 🙏 - I'm happy for this to be reviewed + tested, however I would suggest we hold off a little longer on the merge + release ([internal context](https://linear.app/gitpod/issue/EXP-3#comment-890d1793)).

**Comment by jankeromnes** [comment]
> I'm happy for this to be reviewed + tested

Thanks @loujaybee! 🙌 

> however I would suggest we hold off a little longer on the merge + release ([internal context](https://linear.app/gitpod/issue/EXP-3#comment-890d1793)).

Sorry if I'm missing something, but what is the benefit of leaving a feature broken for a longer time? (I.e. what is the downside of fixing this small bug as soon as possible?)

To me, the slightly broken feature and the larger customer UX discovery are two completely separate topics.

File: components/server/src/bitbucket/bitbucket-repository-provider.ts [hunk 1]
```diff
@@ -59,24 +59,39 @@ export class BitbucketRepositoryProvider implements RepositoryProvider {
...
-                commit: {
-                    sha: branch.target?.hash!,
-                    author: branch.target?.author?.user?.display_name!,
-                    authorAvatarUrl: branch.target?.author?.user?.links?.avatar?.href,
-                    authorDate: branch.target?.date!,
-                    commitMessage: branch.target?.message || "missing commit message",
-                },
+        // Handle pagination.
+        let nextPage: number | undefined = 1;
+        let isMoreDataAvailable: boolean = true;
```
Comment (filiptronicek) [inline components/server/src/bitbucket/bitbucket-repository-provider.ts:65]: I think we can safely remove the type annotations here
```suggestion
        let nextPage = 1;
        let isMoreDataAvailable = true;
```

**Review by filiptronicek — APPROVED** [review]
Code changes LGTM. Did not test because preview environment does not seem to have been triggered 👀

**Comment by jankeromnes** [comment]
Many thanks @filiptronicek for the review! And for pointing out that there is no preview env. I've attempted to get GitHub Actions to produce one. Let's see if it does. 👀

**Comment by filiptronicek** [comment]
It took 2 minutes, but I successfully got all of the branches of my gitpod-io/gitpod mirror back!

image

**Comment by jankeromnes** [comment]
I've removed the unnecessary type annotations and added a `pagelen: 100` parameter, in the hopes of optimizing the number of API round-trips necessary to fetch long lists of branches (if the Bitbucket API behaves like the GitLab and GitHub ones, I believe the default page length might be 30).

**Comment by jankeromnes** [comment]
@filiptronicek Could you please briefly test with your fork again? 😇 🙏 I believe it should be faster now (and still correct!)

File: components/server/src/bitbucket/bitbucket-repository-provider.ts [hunk 2]
```diff
@@ -59,24 +59,40 @@ export class BitbucketRepositoryProvider implements RepositoryProvider {
...
-                    author: branch.target?.author?.user?.display_name!,
-                    authorAvatarUrl: branch.target?.author?.user?.links?.avatar?.href,
-                    authorDate: branch.target?.date!,
-                    commitMessage: branch.target?.message || "missing commit message",
-                },
+        // Handle pagination.
+        let nextPage = 1;
+        let isMoreDataAvailable = true;
+
+        while (isMoreDataAvailable) {
```
Comment (akosyakov) [inline components/server/src/bitbucket/bitbucket-repository-provider.ts:67]: Are there risk that hit some rate limitting here?

Comment (jankeromnes) [reply components/server/src/bitbucket/bitbucket-repository-provider.ts:67]: Good question! 🎯 Probably yes (just like all other Bitbucket API requests), but I think using a `pagelen` of `100` mitigates this somewhat (i.e. you'd have to have an incredibly large number of branches in order to get into rate limit territory, and we also do all these requests in sequence and not in parallel).

My feeling here is that it's safe to let any (supposedly very rare & unlikely) rate limit error just bubble up as is if it ever happens, i.e. the Gitpod user would see something like "Unable to fetch branches: Bitbucket API rate limit exceeded".

Comment (akosyakov) [reply components/server/src/bitbucket/bitbucket-repository-provider.ts:67]: For repositories we have pagination as well but it broke with 10000 as far as I understand and was very slow to fetch everything breaking ability to create a project.

**Review by filiptronicek — APPROVED** [review]
Tested with my mirror again (https://bitbucket.org/biased-configuration/gitpod/src/main/) and it seems to work much faster now. Thanks!

File: components/server/src/bitbucket/bitbucket-repository-provider.ts (hunk 2 above)
Comment (jankeromnes) [reply components/server/src/bitbucket/bitbucket-repository-provider.ts:67]: @akosyakov aha -- was that with `pagelen: 100` or the default page length?

**Comment by jankeromnes** [comment]
@loujaybee good to remove the hold? 🚀

**Comment by loujaybee** [comment]
> Sorry if I'm missing something, but what is the benefit of leaving a feature broken for a longer time? (I.e. what is the downside of fixing this small bug as soon as possible?)

* The context is that this page is planned to be removed, so I'm conscious that we don't set expectations that this page will be invested in more if possible, and I'm also hesitant in introducing more complexity into the page. 
* Yeah, and that was really also my question, e.g. "what is the downside"? If this won't introduce any subsequent other issues, such as performance related issues, exhausting token rate limits, etc then let's go ahead and merge. 

I'm happy to trust your judgement on it, if you think it's a net benefit with minima/no downside, let's merge 🙏

File: components/server/src/bitbucket/bitbucket-repository-provider.ts (hunk 2 above)
Comment (akosyakov) [reply components/server/src/bitbucket/bitbucket-repository-provider.ts:67]: I think it was 10K repos with 1000 page size

**Comment by jankeromnes** [comment]
Many thanks @loujaybee! In this case, I'd like to go ahead with merging this fix:
- It fixes our getBranches feature, which is currently broken for Bitbucket when there are more than ~30 branches
- I believe there aren't any downsides or significant risks (we talked about a possible rate-limit issue, but that can only happen if there are more than 10K or 30K branches in a repo -- quite unlikely)

In any case, if there turns out to be a downside to this fix, I'm happy to revert it and improve my fix. 👍

**Pull request merged**